### PR TITLE
separate branch for pr and review

### PR DESCRIPTION
**Pull request by sojiadeshina**
gluon activations tutorial.

### CODE REVIEW

File: gluon-activations/gluon-activations.md [hunk 1]
```diff
@@ -0,0 +1,239 @@
+
+## MXNet Activations Tutorial
```
Comment (thomelane) [inline gluon-activations/gluon-activations.md]: think we're calling these guides

File: gluon-activations/gluon-activations.md [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+
+Deep neural networks encode a nonlinear, highly parametrized function from input to outputs. The nonlinearities that allow neural networks to be maximally expressive are referred to as activation functions. Over the course of the development of neural networks, several nonlinear activation functions have been introduced to make gradient-based deep learning tractable. 
+
+If you are looking to answer the question, which activation function should I use for my neural network model you should probably go with *ReLU*. Unless, you're trying to implement something like a gating mechanism, like in LSTMs or GRU cells, then you should opt for sigmoid and/or tanh in those cells. However, if you have a working model architecture and you're trying to improve it's performance by swapping out activation functions or treating the activation function as a hyperparameter, then you may want to try hand-designed activations like SELU or RL discovered functions like Swish. This tutorial describes these activation functions and others implemented in mxnet in detail.
```
Comment (thomelane) [inline gluon-activations/gluon-activations.md]: ... question 'which activation function should I use for my neural network model?' you...

Comment (thomelane) [inline gluon-activations/gluon-activations.md]: Unless, -> Unless

Comment (thomelane) [inline gluon-activations/gluon-activations.md]: mxnet -> MXNet

File: gluon-activations/gluon-activations.md [hunk 3]
```diff
@@ -0,0 +1,239 @@
+
+## MXNet Activations Tutorial
+
+Deep neural networks encode a nonlinear, highly parametrized function from input to outputs. The nonlinearities that allow neural networks to be maximally expressive are referred to as activation functions. Over the course of the development of neural networks, several nonlinear activation functions have been introduced to make gradient-based deep learning tractable.
```
Comment (thomelane) [inline gluon-activations/gluon-activations.md]: Sounds a bit to formal/'mathsy' for a beginner. "highly parametrized", "maximally expressive", "learning tractable". See if you can reword a bit. Can get more complex as you go on but start really simple.

File: gluon-activations/gluon-activations.md [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+If you are looking to answer the question, which activation function should I use for my neural network model you should probably go with *ReLU*. Unless, you're trying to implement something like a gating mechanism, like in LSTMs or GRU cells, then you should opt for sigmoid and/or tanh in those cells. However, if you have a working model architecture and you're trying to improve it's performance by swapping out activation functions or treating the activation function as a hyperparameter, then you may want to try hand-designed activations like SELU or RL discovered functions like Swish. This tutorial describes these activation functions and others implemented in mxnet in detail.
+
+### Visualizing Activations
+In order to compare the various activation functions and to understand the nuances of their differences we have a snippet of code to to plot the activation functions and their gradients.
```
Comment (thomelane) [inline gluon-activations/gluon-activations.md]: suggestion: ... plot the activation functions (used in the forward pass) and their gradients (used in the backward pass).

File: gluon-activations/gluon-activations.md [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+### Sigmoid
+The sigmoid activation function, also known as the logistic function or logit function, is perhaps the most widely known activation owing to its [long history](https://web.stanford.edu/class/psych209a/ReadingsByDate/02_06/PDPVolIChapter8.pdf) in neural network training and appearance in logistic regression and kernel methods for classification. 
+
+The sigmoid activation is a non-linear function that transforms any real valued input to a value between 0 and 1, giving it a natural probabilistic interpration. The sigmoid takes the form of the function below
+
+$$ \sigma(x) = \dfrac{1}{1 + e^x} $$ or alternatively
+
+$$ \sigma(x) = \dfrac{e^x}{e^x + 1} $$
+
+Note: the term sigmoid is overloaded and can be used to refer to the class of 's' shaped functions or particularly to the logistic function that we've just described. In MxNet the sigmoid activation specifically refers to logistic function sigmoid.
```
Comment (thomelane) [inline gluon-activations/gluon-activations.md]: Note: -> Warning:
MxNet -> MXNet

File: gluon-activations/gluon-activations.md [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+
+```python
+visualize_activation(data, gluon.nn.Activation('sigmoid'), "Sigmoid")
+```
+
+
+![png](output_5_0.png)
+
+
+The sigmoid activation has since fallen out of use as the preferred activation function in designing neural networks due to some of it's properties, shown in the plot above, like not being zero-centered and inducing vanishing gradients, that leads to poor performance during neural network training. Vanishing gradients here refers to the tendency of the gradient of the sigmoid function to be nearly zero for most input values.
```
Comment (thomelane) [inline gluon-activations/gluon-activations.md]: it's -> its

File: gluon-activations/gluon-activations.md [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+```python
+visualize_activation(data, gluon.nn.Activation('tanh'), "Tanh")
+```
+
+
+![png](output_8_0.png)
+
+
+The use tanh as activation functions in place of the logistic function was popularized by the success of the [LeNet architecture](http://yann.lecun.com/exdb/publis/pdf/lecun-01a.pdf) and the [methods paper](http://yann.lecun.com/exdb/publis/pdf/lecun-98b.pdf) by LeCun et al
```
Comment (thomelane) [inline gluon-activations/gluon-activations.md]: The use tanh -> The use of tanh

File: gluon-activations/gluon-activations.md [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+```
+
+
+![png](output_5_0.png)
+
+
+The sigmoid activation has since fallen out of use as the preferred activation function in designing neural networks due to some of it's properties, shown in the plot above, like not being zero-centered and inducing vanishing gradients, that leads to poor performance during neural network training. Vanishing gradients here refers to the tendency of the gradient of the sigmoid function to be nearly zero for most input values. 
+
+### tanh
+The tanh activation function is also an s shaped curve defined by the mathematical equation:
```
Comment (thomelane) [inline gluon-activations/gluon-activations.md]: Can you give an example of when to use before diving into the maths?

File: gluon-activations/gluon-activations.md [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+
+
+![png](output_8_0.png)
+
+
+The use tanh as activation functions in place of the logistic function was popularized by the success of the [LeNet architecture](http://yann.lecun.com/exdb/publis/pdf/lecun-01a.pdf) and the [methods paper](http://yann.lecun.com/exdb/publis/pdf/lecun-98b.pdf) by LeCun et al
+
+### SoftSign
+
+The softsign activation is an alternative to tanh that is also centered at zero but converges asymptotically to -1 and 1 polynomially instead of exponentially. As such, there are a greater range of input values for which the softsign assigns an output of between -1 and 1.
```
Comment (thomelane) [inline gluon-activations/gluon-activations.md]: As such, there are a greater range of input values for which the softsign assigns an output of between -1 and 1.
A bit confused? It's the same isn't it?

File: gluon-activations/gluon-activations.md [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+
+```python
+visualize_activation(data, gluon.nn.Activation('softsign'), "SoftSign")
+```
+
+
+![png](output_11_0.png)
+
+
+The softsign function is not a commonly used activation with most neural networks and suffers even more severely from the vanishing gradient problem as seen in the graph above
```
Comment (thomelane) [inline gluon-activations/gluon-activations.md]: It looks like less of an issue to me!
Also add . at end.

File: gluon-activations/gluon-activations.md [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+
+## Rectifiers
+
+### ReLU
+ReLU, or Rectified linear unit is the most common activation function in convolutional neural networks and introduces a simple nonlinearity. When the value of the input into ReLU is positive, then it retains the same value. When the value is negative then it becomes zero. In equation form, the ReLU function is given as:
```
Comment (thomelane) [inline gluon-activations/gluon-activations.md]: Rectified linear unit -> Rectified Linear Unit

File: gluon-activations/gluon-activations.md [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+
+```python
+visualize_activation(data, gluon.nn.Activation('relu'), "ReLU")
+```
+
+
+![png](output_15_0.png)
+
+
+As shown above, the ReLU activation ameliorates the vanishing gradient problem associated with the sigmoid family of activations, by having a larger (infinite) range of values where it's gradient is non-zero.
```
Comment (thomelane) [inline gluon-activations/gluon-activations.md]: 'ameliorates' 😮 might want to simplify for us illiterate folk.

File: gluon-activations/gluon-activations.md [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+
+### SoftReLU
+
+SoftReLU also known as SmoothReLU or SoftPlus is a nonlinear activation function that takes the form
+
+$$ softReLU(x) = log(1 + e^x)$$
+
+The softReLU can be seen as a smooth version of the ReLU by observing that it's derivative is the sigmoid, seen below, which is a smooth version of the gradient of the ReLU shown above
```
Comment (thomelane) [inline gluon-activations/gluon-activations.md]: softReLU -> SoftReLU

Comment (thomelane) [reply gluon-activations/gluon-activations.md]: . at end

File: gluon-activations/gluon-activations.md [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+### Leaky ReLU
+
+Leaky ReLUs are a variant of ReLU that allow a small gradient when the value is negative. The equation for the LeakyReLU is:
+
+$$ LeakyReLU(\alpha, x) = \begin{cases}
+    x,& \text{if } x\geq 0\\
+    \alpha x,              & \text{otherwise}
+\end{cases}$$ 
+
+where $\alpha > 0$ is small positive number. Here is a visualization for the LeakyReLU with $\alpha = 0.01$
```
Comment (thomelane) [inline gluon-activations/gluon-activations.md]: Any chance you could exaggerate alpha for the diagrams, but mention 0.01 is most common (if it actually is).

File: gluon-activations/gluon-activations.md [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+
+
+![png](output_11_0.png)
+
+
+The softsign function is not a commonly used activation with most neural networks and suffers even more severely from the vanishing gradient problem as seen in the graph above
+
+## Rectifiers
+
+### ReLU
```
Comment (thomelane) [inline gluon-activations/gluon-activations.md:106]: Add a warning in this section about dying ReLUs

File: gluon-activations/gluon-activations.md [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+
+
+### Swish
+Swish is an activation function that attempts to address the shortcomings of ReLU by combining ideas from ReLU and sigmoid. Swish was discovered by searching the space of activation functions using a combination of exhaustive and reinforcement learning-based search and was introduced in the paper by [Ramchandran et al](https://arxiv.org/pdf/1710.05941.pdf)
+
+The swish function is given as 
+
+$$ swish(x) = x\cdot\sigma(\beta x)$$
+
+where $\sigma$ is the sigmoid activation function $\sigma(x) = \frac{1}{1 + e^{-x}}$ described above and $\beta$ is a hyperparameter set to 1 by default in mxnet
```
Comment (thomelane) [inline gluon-activations/gluon-activations.md]: mxnet -> MXNet

File: gluon-activations/gluon-activations.md (hunk 2 above)
Comment (thomelane) [inline gluon-activations/gluon-activations.md]: RL -> reinforcement learning

**Review by thomelane — COMMENTED** [review]
Great overview! I learnt a load.
Just wanted a few more examples of networks using each of these activations (which might be hard to find) and why.
And on occasions a bit more intuition before diving into the formulas.

File: gluon-activations/gluon-activations.md (hunk 2 above)
Comment (sojiadeshina) [reply gluon-activations/gluon-activations.md]: thanks for catching

Comment (sojiadeshina) [reply gluon-activations/gluon-activations.md]: will address

Comment (sojiadeshina) [reply gluon-activations/gluon-activations.md]: thanks for catching

File: gluon-activations/gluon-activations.md (hunk 3 above)
Comment (sojiadeshina) [reply gluon-activations/gluon-activations.md]: How about:

"Deep neural networks are a way to express a nonlinear function with lots of parameters from input data to outputs. The nonlinearities that allow neural networks to capture complex patterns in data are referred to as activation functions."

File: gluon-activations/gluon-activations.md (hunk 4 above)
Comment (sojiadeshina) [reply gluon-activations/gluon-activations.md]: i like it. thanks!

File: gluon-activations/gluon-activations.md (hunk 8 above)
Comment (sojiadeshina) [reply gluon-activations/gluon-activations.md]: okay. will add example

File: gluon-activations/gluon-activations.md (hunk 9 above)
Comment (sojiadeshina) [reply gluon-activations/gluon-activations.md]: yeah i rephrased this afterwards. what i mean is that softsign doesn't saturate as quickly as tanh. i'll find another way to say this in lay terms

File: gluon-activations/gluon-activations.md (hunk 10 above)
Comment (sojiadeshina) [reply gluon-activations/gluon-activations.md]: i guess the answer is it depends. the range of values to large (closer to one) gradients is narrower for softsign but the range of values closer to zero is larger for tanh. i'm gonna reword this though since to decide on way or the other depends on initialization

File: gluon-activations/gluon-activations.md (hunk 11 above)
Comment (sojiadeshina) [reply gluon-activations/gluon-activations.md]: thanks

File: gluon-activations/gluon-activations.md (hunk 15 above)
Comment (sojiadeshina) [reply gluon-activations/gluon-activations.md:106]: added. thanks

File: gluon-activations/gluon-activations.md (hunk 16 above)
Comment (sojiadeshina) [reply gluon-activations/gluon-activations.md]: changed. Thanks

File: gluon-activations/gluon-activations.md (hunk 3 above)
Comment (thomelane) [reply gluon-activations/gluon-activations.md]: 👍

**Comment by sojiadeshina** [comment]
incorporated feedback. gonna submit pr on new-docs fork now
